diff --git a/cyclone_objects/binaries/control/coll.c b/cyclone_objects/binaries/control/coll.c
index 3926b58e..bf8dd5da 100644
--- a/cyclone_objects/binaries/control/coll.c
+++ b/cyclone_objects/binaries/control/coll.c
@@ -29,8 +29,9 @@ before used to always bang on callback, now only bangs now due to new x->x_fileb
 //#define COLL_DEBUG
 #endif */
 
-
+#ifndef COLLTHREAD
 #define COLLTHREAD 1 //default is threaded
+#endif
 #define COLLEMBED 0 //default for save in patch
 #define COLL_ALLBANG 1 //bang all when read instead of specific object
 #define COLL_MAXEXTLEN 4 //maximum file extension length
